### ISSUE TITLE
eslint + babel-eslint + ArrowFunctionExpression as class method = inaccurate getJSDocComment

### ISSUE THREAD

**Issue by brokentone** [eslint/eslint#10790]
https://github.com/brokentone/eslint/tree/bug/babel-eslint-arrow-jsdoc
**Tell us about your environment**

* **ESLint Version:** master
* **Node Version:** v8.11.3
* **npm Version:** 5.6.0

**What parser (default, Babel-ESLint, etc.) are you using?** babel-eslint

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{ rules: { checker: "error" }, parser: "babel-eslint", parserOptions: { ecmaVersion: 6, ecmaFeatures: { jsx: true } } }
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
/** Merges two objects together.*/
class A {
    constructor(xs) {}

    methodOne = () => {}
};
```

```bash
./node_modules/.bin/mocha tests/lib/util/source-code.js
```

**What did you expect to happen?**
Not to associate the comment with the arrow function with `sourceCode.getJSDocComment(node);`

**What actually happened? Please include the actual, raw output from ESLint.**
Comment returned.

I have a test case built here: https://github.com/brokentone/eslint/tree/bug/babel-eslint-arrow-jsdoc

I added the `babel-eslint` parser to all the other rules in the file as a second commit, and the only violation is this new case I added.

I attempted to dig through `getJSDocComment` and I couldn't quite figure out the issue. In fairness, it may very well be a `babel-eslint` issue.

**Comment by brokentone** [comment]
So I have figured out that the AST for my demo case looks something like so: https://astexplorer.net/#/gist/dde17821e86a0fda5ffc7e9f1875e30b/2b13f08d27748b8c081e0849bd267b269582d8f9

Meaning from my node, walking back up the tree to roughly where my comment node is, we have: `ArrowFunctionExpression => ClassProperty => ClassBody => ClassDeclaration`

None of these nodes walk up the parents in `getJSDocComment` around here: https://github.com/eslint/eslint/blob/master/lib/util/source-code.js#L361

If, however, you add `ClassProperty`, then my issue seems to be resolved, yay! More trouble. The rules say that we cannot use this feature:

> ESLint's parser only officially supports the latest final ECMAScript standard. We will make changes to core rules in order to avoid crashes on stage 3 ECMAScript syntax proposals (as long as they are implemented using the correct experimental ESTree syntax). We may make changes to core rules to better work with language extensions (such as JSX, Flow, and TypeScript) on a case-by-case basis.

It appears now that there are some paths. We could...
1. Suggest to `babel-eslint` that they patch the whole `getJSDocComment` util -- but since it doesn't seem patched yet, this is likely a ton of work and maintenance to ask them to do for ultimately a single line change.
2. Could loosen the `parent.type !== "Property"` condition to be like function above it: `!/Property/.test(parent.type)` -- this gets us a bit future proofed, and doesn't lock in an experimental token name.
3. Alternately could replace `parent.type !== "MethodDefinition"` (and possibly `parent.type !== "Property"`) with `parent.type !== "ClassBody"`. This is a token further away, but it seems like a reasonable structure to me.

Points 2 and 3 both solve my issue, and can be easily addressed here. Would you mind considering either of them? I'm happy to write the (one line) of code.

**Comment by platinumazure** [comment]
I don't know if the rest of the team feels the same way necessarily, but I would be open to a change to `getJSDocComment` for your use case. In my view, the intention of the rule policy is that we want users to create custom rules for experimental language features. But in this case, a core API is being hamstrung by this limitation and so writing a custom rule is more difficult than it should be for this case. So I'm hoping we can do something here.

That said, the team may ultimately suggest a different solution.

**Comment by kaicataldo** [comment]
I would be for making a change that would fix this case, but I want to make sure it's a generic solution that isn't tied to a particular parser.

**Comment by brokentone** [comment]
Thanks for the guidance @platinumazure and @kaicataldo! Do you think this approach is sufficiently generic? https://github.com/eslint/eslint/pull/10792

**Comment by nzakas** [comment]
Thanks for the suggestion. As of yesterday, we have decided to officially [end-of-life JSDoc support](https://eslint.org/blog/2018/11/jsdoc-end-of-life) in ESLint. All JSDoc features are now deprecated and we won't be fixing bugs or making any further improvements to those features.

We are recommending that people transition over to use the [`eslint-plugin-jsdoc`](https://github.com/gajus/eslint-plugin-jsdoc) plugin instead of the built-in rules in ESLint.

Thanks for understanding and we apologize for any inconvenience.
